node/core/dispute-coordinator/src/participation/mod.rs

```diff
@@ -319,6 +323,23 @@ async fn participate(
 		},
 	};
 
+	let (executor_params_tx, executor_params_rx) = oneshot::channel();
+	sender
+		.send_message(RuntimeApiMessage::Request(
+			req.candidate_receipt().descriptor.relay_parent,
+			RuntimeApiRequest::SessionExecutorParams(executor_params_tx),
+		))
+		.await;
+
+	let executor_params = match executor_params_rx.await {
+		Err(_) | Ok(Err(_)) => ExecutorParams::default(), // Runtime API is not yet available
```

Comment:
> Between runtime API upgrade and new session --- can be fixed by my comment above.

Why would it return a runtime API error in that case?

I believe [this call](https://github.com/paritytech/polkadot/pull/6161/files#diff-5be50bcf911f9f87886996cbd93a14eb24a6d2fc51a036972caa0df378ce6883R33-R36) would always return `Ok(None)`, not `Err`, if the value is absent in [storage map](https://github.com/paritytech/polkadot/pull/6161/files#diff-f5965527c1800028825b35b5cd1cc225f3f52edecb006d4c43c91bc2a32b2462R120-R123)

Reply:
Yes, you're right. It was a late-night comment :sweat_smile:

Reply:
> One thing that could help here is caching execution params in runtime API cache not by `relay_parent`, but by `session_index` (see how this is done for the `session_info` runtime call). In that case, request will need to supply `session_index`.

Well, the whole story behind such an implementation of the runtime API request (from relay parent hash to executor params) is about not having a session index on hands. When a candidate is to be validated, we only have `CandidateDescriptor` which only contains relay parent hash, but not a session index. Solutions discussed were:
* To add `session_index` field to the `CandidateDescriptior` struct, but it turned out to require a rather complex migration, and a solution requesting by `relay_parent` seemed much cheaper;
* To add a separate mapping from `relay_parent` to `session_index`, but @eskimor has convinced us that is not needed as we can recover the session index from a state. I remember a discussion on pruned states also took place, and you guys somehow concluded that it should work anyway (I was listening and trying to catch up with little luck).

So I'm not really sure what has to be done here. To me it sounds like we don't really care, as if the block has already been finalized, a dispute outcome doesn't matter any more. But I suppose I'm still missing some parts of a big picture.

Reply:
Let me elaborate on my thinking how it could help:

The scenario when the runtime API would return an error about pruning is the following:
1. We import a block with the given relay parent.
2. We call a bunch of runtime API calls on it right away (but not `session_executor_params` -- because it's only called when doing backing or approval-voting). All calls are now cached, including `session_index_for_child` and runtime version -- these methods should be queried on every block import in may subsystems when seeing a new leaf.
3. A block at the same height was finalized - this block state is now pruned.
4. What we can do is call `session_index_for_child` on that block - it should go directly to the cache even if the block is pruned.
5. And with the `session_index` retrieved from the cache we can retrieve a cached `session_executor_params`.

Let me know if this sounds too contrived.

Reply:
> To me it sounds like we don't really care, as if the block has already been finalized, a dispute outcome doesn't matter any more.

I *think* we want a dispute to conclude even if there's a race with finality so we can slash any attempt of an attacker. The whole point of slashing is to prevent repeated attempts at attacks ([gambler's ruin](https://en.wikipedia.org/wiki/Gambler%27s_ruin)).